core/src/main/clojure/xtdb/expression/temporal.clj

```diff
@@ -460,6 +460,107 @@
+   :->call-code (fn [[x]]
+                  (case start-field
+                    "YEAR" `(PeriodDuration. (Period/ofYears ~x) Duration/ZERO)
+                    "MONTH" `(PeriodDuration. (Period/ofMonths ~x) Duration/ZERO)
+                    "DAY" `(PeriodDuration. (Period/ofDays ~x) Duration/ZERO)
+                    "HOUR" `(PeriodDuration. Period/ZERO (Duration/ofHours ~x))
+                    "MINUTE" `(PeriodDuration. Period/ZERO (Duration/ofMinutes ~x))
+                    "SECOND" `(PeriodDuration. Period/ZERO (Duration/ofSeconds ~x))))})
+
+(defn ->single-field-interval-call [{{:keys [start-field leading-precision fractional-precision]} :cast-opts}]
+  (let [expr {:op :call
+              :f :single_field_interval
+              :args [{} {:literal start-field} {:literal leading-precision} {:literal fractional-precision}]
+              :arg-types [:utf8 :utf8 :int :int]}]
+    (expr/codegen-call expr)))
+
+(defn ->multi-field-interval-call [{{:keys [start-field end-field leading-precision fractional-precision]} :cast-opts}]
+  (let [expr {:op :call
+              :f :multi_field_interval
+              :args [{} {:literal start-field} {:literal leading-precision} {:literal end-field} {:literal fractional-precision}]
+              :arg-types [:utf8 :utf8 :int :utf8 :int]}]
+    (expr/codegen-call expr)))
+
+(defmethod expr/codegen-cast [:utf8 :interval] [{{:keys [start-field end-field]} :cast-opts :as expr}]
+  ;; Calls down to multi-field-interval if both start-field and end-field present, 
+  ;; else calls to single-field-interval if only start-field is present
+  (if (and start-field end-field)
+    (->multi-field-interval-call expr)
+    (->single-field-interval-call expr)))
+
+(defn ym-interval->string [^PeriodDuration x]
+  (let [period (.getPeriod x)
+        sign (if (.isNegative period) "-" "")
+        abs-total-months (abs (.toTotalMonths period))
+        year (quot abs-total-months 12)
+        month (mod abs-total-months 12)
+        interval-string (cond-> ""
+                          (not= year 0) (str year (when (not= month 0) "-"))
+                          (not= month 0) (str month))
+        interval-qualifier (cond-> ""
+                             (not= year 0) (str "YEAR" (when (not= month 0) " TO "))
+                             (not= month 0) (str "MONTH"))]
+    (format "INTERVAL '%s%s' %s" sign interval-string interval-qualifier)))
+
+(defn- get-present-fields [days hours minutes seconds nanos]
+  (cond-> []
+    (not= days 0) (conj {:field "DAY" :rank 1})
+    (not= hours 0) (conj {:field "HOUR" :rank 2})
+    (not= minutes 0) (conj {:field "MINUTE" :rank 3})
+    (not= seconds nanos 0) (conj {:field "SECOND" :rank 4})))
+
+(defn- format-end-field [start-field end-field nanos?]
+  (cond
+    (and (= start-field end-field) nanos?) "SECOND(1,9)"
+    nanos? "SECOND(9)"
+    :else end-field))
+
+(defn- ->interval-str [days hours minutes seconds nanos start-rank end-rank]
```

Comment:
this is likely where you're getting your boxed-math warnings - all of the numeric parameters get boxed up here to call this function. 

unfortunately we can't put primitive type-hints on functions with more than 4 parameters in Clojure - possible to take the period and duration in here instead, perhaps?